### ISSUE TITLE
High Availability Concepts and Best Practices

### ISSUE THREAD

**Issue by guevara** [guevara/read-it-later#6447]
High Availability Concepts and Best Practices<br>
<br>
https://ift.tt/2bFOzsD<br>
<br>
About 0 Minutes<br>
<br>
<div>
<div>
<a name="13572" id="13572"></a> <a name="RACCN010" id="RACCN010"></a><a name="13303" id="13303"></a>
<p>This chapter describes the concepts and some of the <em>best practices</em> methodologies for use in Real Application Clusters to implement high availability. This chapter includes the following topics:</p>
<ul>
<li type="disc">
<a name="6056" id="6056"></a><a href="https://docs.oracle.com/cd/A91202_01/901_doc/rac.901/a89867/pshavdtl.htm#6062">Understanding High Availability</a>
</li>
<li type="disc">
<a name="6057" id="6057"></a><a href="https://docs.oracle.com/cd/A91202_01/901_doc/rac.901/a89867/pshavdtl.htm#6913">Planning for High Availability</a>
</li>
<li type="disc">
<a name="6058" id="6058"></a><a href="https://docs.oracle.com/cd/A91202_01/901_doc/rac.901/a89867/pshavdtl.htm#9272">Configuring Real Application Clusters for High Availability</a>
</li>
<li type="disc">
<a name="15544" id="15544"></a><a href="https://docs.oracle.com/cd/A91202_01/901_doc/rac.901/a89867/pshavdtl.htm#6941">Disaster Planning</a>
</li>
<li type="disc">
<a name="10765" id="10765"></a><a href="https://docs.oracle.com/cd/A91202_01/901_doc/rac.901/a89867/pshavdtl.htm#6945">Failure Protection Validation</a>
</li>
<li type="disc">
<a name="10766" id="10766"></a><a href="https://docs.oracle.com/cd/A91202_01/901_doc/rac.901/a89867/pshavdtl.htm#13837">Failover and Real Application Clusters</a>
</li>
<li type="disc">
<a name="15549" id="15549"></a><a href="https://docs.oracle.com/cd/A91202_01/901_doc/rac.901/a89867/pshavdtl.htm#9371">Failover Basics</a>
</li>
<li type="disc">
<a name="15553" id="15553"></a><a href="https://docs.oracle.com/cd/A91202_01/901_doc/rac.901/a89867/pshavdtl.htm#10853">Client Failover</a>
</li>
<li type="disc">
<a name="15557" id="15557"></a><a href="https://docs.oracle.com/cd/A91202_01/901_doc/rac.901/a89867/pshavdtl.htm#7011">Server Failover</a>
</li>
<li type="disc">
<a name="15561" id="15561"></a><a href="https://docs.oracle.com/cd/A91202_01/901_doc/rac.901/a89867/pshavdtl.htm#7013">Host-Based Failover</a>
</li>
<li type="disc">
<a name="15566" id="15566"></a><a href="https://docs.oracle.com/cd/A91202_01/901_doc/rac.901/a89867/pshavdtl.htm#13867">Real Application Clusters Failover</a>
</li>
<li type="disc">
<a name="15571" id="15571"></a><a href="https://docs.oracle.com/cd/A91202_01/901_doc/rac.901/a89867/pshavdtl.htm#15337">How Failover Works</a>
</li>
<li type="disc">
<a name="10779" id="10779"></a><a href="https://docs.oracle.com/cd/A91202_01/901_doc/rac.901/a89867/pshavdtl.htm#10874">High Availability Configurations</a>
</li>
<li type="disc">
<a name="10784" id="10784"></a><a href="https://docs.oracle.com/cd/A91202_01/901_doc/rac.901/a89867/pshavdtl.htm#10556">Toward Deploying High Availability</a>
</li>
</ul>
<a name="6062" id="6062"></a>
<h2><span color="#330099">Understanding High Availability</span></h2>
<a name="6575" id="6575"></a>
<p>Computing environments configured to provide nearly full-time availability are known as <a href="https://docs.oracle.com/cd/A91202_01/901_doc/rac.901/a89867/glossary.htm#432506"><strong>high availability</strong></a> systems. Such systems typically have redundant hardware and software that makes the system available despite failures. Well-designed high availability systems avoid having single points-of-failure. Any hardware or software component that can fail has a redundant component of the same type.</p>
<a name="6092" id="6092"></a>
<p>When failures occur, the <a href="https://docs.oracle.com/cd/A91202_01/901_doc/rac.901/a89867/glossary.htm#432337"><strong>failover</strong></a> process moves processing performed by the failed component to the backup component. This process remasters systemwide resources, recovers partial or failed transactions, and restores the system to normal, preferably within a matter of microseconds. The more transparent that failover is to users, the higher the availability of the system.</p>
<a name="14952" id="14952"></a>
<p>Oracle has a number of products and features that provide high availability. These include Real Application Clusters, <a href="https://docs.oracle.com/cd/A91202_01/901_doc/rac.901/a89867/glossary.htm#434008"><strong>Oracle Real Application Clusters Guard</strong></a>, Oracle Replication, and <a href="https://docs.oracle.com/cd/A91202_01/901_doc/rac.901/a89867/glossary.htm#432886"><strong>Oracle9i Data Guard</strong></a>. These can be used in various combinations to meet your specific high availability needs. This chapter describes high availability within the context of Real Application Clusters.</p>
<a name="15581" id="15581"></a>
<p>Real Application Clusters are inherently high availability systems. Clusters typical of Real Application Clusters environments, as described in <a href="https://docs.oracle.com/cd/A91202_01/901_doc/rac.901/a89867/pshwarch.htm#6175">Chapter&nbsp;3</a>, can provide continuous service for both planned and unplanned outages.</p>
<a name="14816" id="14816">
<div align="center">
<span><strong>Note:</strong></span><a name="14827" id="14827"></a>
<p>More on the topic of High Availability is included in</p>
<a href="https://docs.oracle.com/cd/A91202_01/901_doc/rac.901/a89867/opfsarch.htm#41868">Chapter&nbsp;11</a>and <a href="https://docs.oracle.com/cd/A91202_01/901_doc/rac.901/a89867/opfsoper.htm#43583">Chapter&nbsp;12</a>. These chapters describe high availability within the context of Oracle Real Application Clusters Guard. &nbsp;
<hr>
<table border="0" cellpadding="0" width="80%" cellspacing="0" dir="ltr">
<tbody>
<tr>
<td>
<hr>
</td>
</tr>
</tbody>
</table>
</div>
</a> <a name="14817" id="14817"></a>
<h3><span color="#330099">Measuring Availability</span></h3>
<a name="6614" id="6614"></a>
<p>You can classify systems and evaluate their expected availability by system type. Mission-critical and business-critical applications such as e-mail and Internet servers probably require a significantly greater availability than do applications that have a smaller number of users. As well, some systems could have a <em>continuous</em> (24 hours a day, seven days a week) uptime requirement, while others such as a stock market tracking system will have nearly continuous uptime requirements for specific time frames, such as when a stock market is open.</p>
<a name="6678" id="6678"></a>
<h3><span color="#330099">High Availability Measurements</span></h3>
<a name="6679" id="6679"></a>
<p>The software industry generally measures availability by using two types of metrics (measurements):</p>
<ul>
<li type="disc">
<a name="6680" id="6680"></a>The mean time to recover</li>
<li type="disc">
<a name="6681" id="6681"></a>The mean time between failures</li>
</ul>
<a name="6682" id="6682"></a>
<p>For most failure scenarios, the industry focuses on <a href="https://docs.oracle.com/cd/A91202_01/901_doc/rac.901/a89867/glossary.htm#432763"><strong>mean time to recover (MTTR)</strong></a> issues and investigates how to optimally design systems to reduce these. MTBF is generally more applicable to hardware availability metrics; this chapter does not go into detail about <a href="https://docs.oracle.com/cd/A91202_01/901_doc/rac.901/a89867/glossary.htm#432757"><strong>mean time between failures (MTBF)</strong></a>. However, given that you can design Real Application Clusters to avoid single points-of-failure, component failures might not necessarily result in application unavailability. Hence, Real Application Clusters can greatly reduce the MTBF from an application availability standpoint.</p>
<a name="6683" id="6683"></a>
<p>Another metric that is generally used is <em>number of nines</em>. For example, 526 minutes of system unavailability for each year results in 99.9% or <a href="https://docs.oracle.com/cd/A91202_01/901_doc/rac.901/a89867/glossary.htm#433739"><strong>three nines availability</strong></a>.</p>
<a name="15517" id="15517"></a>
<p>Five minutes of system unavailability for each year results in a 99.999% or <a href="https://docs.oracle.com/cd/A91202_01/901_doc/rac.901/a89867/glossary.htm#432360"><strong>five nines availability</strong></a>.</p>
<a name="6684" id="6684"></a>
<p>It is difficult to consider several nines of availability without also describing the ground rules and strict processes for managing application environments, testing methodologies, and change management procedures. For these reasons, Real Application Clusters can significantly reduce MTTR during failures. This inevitably contributes toward a more favorable availability for an entire system.</p>
<a name="9642" id="9642"></a>
<p>As mentioned, a well designed Real Application Clusters system has redundant components that protect against most failures and that provide an environment without single points-of-failure. Working with your hardware vendor is key to building fully redundant cluster environments for Real Application Clusters.</p>
<a name="15033" id="15033"></a>
<h3><span color="#330099">Causes of Downtime</span></h3>
<a name="15037" id="15037"></a>
<p>Downtime can be either planned or unplanned. <a href="https://docs.oracle.com/cd/A91202_01/901_doc/rac.901/a89867/pshavdtl.htm#15041">Figure&nbsp;10-1</a> shows many causes of <a href="https://docs.oracle.com/cd/A91202_01/901_doc/rac.901/a89867/glossary.htm#433793"><strong>unplanned downtime</strong></a>. They can be broadly classified as system faults, data and media errors, and site outages. This unplanned downtime is disruptive because it is difficult to predict its timing.</p>
<a name="15038" id="15038"></a> <a name="15039" id="15039"></a> <a name="15041" id="15041"></a>
<h4><span><em>Figure 10-1 Causes of Unplanned Downtime</em></span></h4>
<a name="15045" id="15045"><img width="600" alt="Text description of pfscn012.gif follows" src="https://docs.oracle.com/cd/A91202_01/901_doc/rac.901/a89867/pfscn012.gif" height="601"></a><a href="https://docs.oracle.com/cd/A91202_01/901_doc/rac.901/a89867/img_text/pfscn012.htm"><br>
Text description of the illustration pfscn012.gif</a> <a name="15046" id="15046"></a> <a name="15047" id="15047"></a>
<p>However, <a href="https://docs.oracle.com/cd/A91202_01/901_doc/rac.901/a89867/glossary.htm#433113"><strong>planned downtime</strong></a> can be just as disruptive to operations, especially in global enterprises that support users in multiple time zones. <a href="https://docs.oracle.com/cd/A91202_01/901_doc/rac.901/a89867/pshavdtl.htm#15053">Figure&nbsp;10-2</a> shows several kinds of planned downtime. They can be generally classified as routine operations, periodic maintenance, and upgrades.</p>
<a name="15051" id="15051"></a> <a name="15053" id="15053"></a>
<h4><span><em>Figure 10-2 Causes of Planned Downtime</em></span></h4>
<a name="15057" id="15057"><img width="600" alt="Text description of pfscn013.gif follows" src="https://docs.oracle.com/cd/A91202_01/901_doc/rac.901/a89867/pfscn013.gif" height="355"></a><a href="https://docs.oracle.com/cd/A91202_01/901_doc/rac.901/a89867/img_text/pfscn013.htm"><br>
Text description of the illustration pfscn013.gif</a> <a name="15058" id="15058"></a> <a name="15059" id="15059"></a>
<p>The key to building a high availability solution is to consider all these causes of downtime and to design a solution that addresses them. Oracle Real Application Clusters Guard is Oracle's premier high availability solution. It has been designed to address many of the causes of unplanned and planned downtime, to maximize database availability, and to minimize disruption to end users of the data services.</p>
<a name="15073" id="15073"></a> <a name="6913" id="6913"></a>
<h2><span color="#330099">Planning for High Availability</span></h2>
<a name="6914" id="6914"></a>
<p>High availability is the result of thorough planning and careful system design. You can conduct high availability planning at two levels:</p>
<ul>
<li type="disc">
<a name="7703" id="7703"></a>The system level with a broad perspective</li>
<li type="disc">
<a name="7704" id="7704"></a>The failure protection level to ensure against a long list of potential causes of failures</li>
</ul>
<a name="6915" id="6915"></a>
<h3><span color="#330099">System Level Planning</span></h3>
<a name="7064" id="7064"></a>
<p>System level planning involves:</p>
<ul>
<li type="disc">
<a name="7065" id="7065"></a><a href="https://docs.oracle.com/cd/A91202_01/901_doc/rac.901/a89867/pshavdtl.htm#6916">Capacity Planning</a>
</li>
<li type="disc">
<a name="7070" id="7070"></a><a href="https://docs.oracle.com/cd/A91202_01/901_doc/rac.901/a89867/pshavdtl.htm#6920">Redundancy Planning</a>
</li>
</ul>
<a name="6916" id="6916"></a>
<h4><span color="#330099">Capacity Planning</span></h4>
<a name="6917" id="6917"></a>
<p>High availability requires the timely processing of transactions for a system to be deemed completely <em>available</em>. While this chapter does not provide extended capacity planning discussions, adequate system resources for managing application growth are important for maintaining availability.</p>
<a name="6918" id="6918"></a>
<p>If an application runs on a single <a href="https://docs.oracle.com/cd/A91202_01/901_doc/rac.901/a89867/glossary.htm#433620"><strong>Symmetric Multi-Processor (SMP)</strong></a> machine with single <a href="https://docs.oracle.com/cd/A91202_01/901_doc/rac.901/a89867/glossary.htm#432546"><strong>instance</strong></a> Oracle, a natural growth path is to migrate this database to a larger SMP machine. However, depending on your hardware vendor's product line, this might not be an option. Migrating to Real Application Clusters, however, might be an option.</p>
<a name="6919" id="6919"></a>
<p>Real Application Clusters enables you to add nodes to your system to increase capacity and handle application growth. You can do this with minimal interference to existing client transactions.</p>
<a name="6920" id="6920"></a>
<h4><span color="#330099">Redundancy Planning</span></h4>
<a name="6921" id="6921"></a>
<p>Redundancy planning means duplicating system components so that no single component failure increases system downtime. Redundant components are often used in high-end SMP machines to protect against failures. For example, redundant power supplies and redundant cooling fans are not uncommon in high-end SMP server systems. A clustered Real Application Clusters environment can extend this redundant architecture by creating complete redundancy so that there is no single point-of-failure.</p>
<a name="9653" id="9653">
<div align="center">
<span><strong><strong>Note</strong>:</strong></span><a name="9658" id="9658"></a>
<p>When installing a high availability system, ensure the hardware and software are certified as a unit. &nbsp;</p>
<hr>
<table border="0" cellpadding="0" width="80%" cellspacing="0" dir="ltr">
<tbody>
<tr>
<td>
<hr>
</td>
</tr>
</tbody>
</table>
</div>
</a> <a name="9272" id="9272"></a>
<h2><span color="#330099">Configuring Real Application Clusters for High Availability</span></h2>
<a name="9273" id="9273"></a>
<p>Real Application Clusters builds higher levels of availability on top of the standard Oracle features. All single instance high availability features such as Fast-Start Recovery and online reorganizations apply to Real Application Clusters as well. Fast-Start Recovery can greatly reduce MTTR with minimal effects on online application performance. Online reorganizations reduce the durations of planned downtimes. Many operations can be performed online while users update the underlying objects. Real Application Clusters preserves all these standard Oracle features.</p>
<a name="9274" id="9274"></a>
<p>In addition to all the regular Oracle features, Real Application Clusters exploits the redundancy provided by clustering to deliver availability with <em>n</em>-1 node failures in an <em>n</em>-node cluster. In other words, all users have access to all data as long as there is one available node in the cluster.</p>
<a name="9295" id="9295"></a>
<p>To configure Real Application Clusters for high availability, you must carefully consider the hardware and software component issues of your cluster as described the following section.</p>
<a name="6922" id="6922"></a>
<h3><span color="#330099">Cluster Components and High Availability</span></h3>
<a name="7242" id="7242"></a>
<p>This section describes high availability and cluster components in the following sections:</p>
<ul>
<li type="disc">
<a name="6923" id="6923"></a><a href="https://docs.oracle.com/cd/A91202_01/901_doc/rac.901/a89867/pshavdtl.htm#7192">Cluster Nodes</a>
</li>
<li type="disc">
<a name="6924" id="6924"></a><a href="https://docs.oracle.com/cd/A91202_01/901_doc/rac.901/a89867/pshavdtl.htm#6930">Cluster Interconnects</a>
</li>
<li type="disc">
<a name="6925" id="6925"></a><a href="https://docs.oracle.com/cd/A91202_01/901_doc/rac.901/a89867/pshavdtl.htm#6933">Storage Devices</a>
</li>
<li type="disc">
<a name="6926" id="6926"></a><a href="https://docs.oracle.com/cd/A91202_01/901_doc/rac.901/a89867/pshavdtl.htm#6936">Operating System Software and Cluster Managers</a>
</li>
<li type="disc">
<a name="6927" id="6927"></a><a href="https://docs.oracle.com/cd/A91202_01/901_doc/rac.901/a89867/pshavdtl.htm#6939">Database Software</a> <a name="9717" id="9717"></a>
</li>
</ul>
<a name="7192" id="7192"></a>
<h4><span color="#330099">Cluster Nodes</span></h4>
<a name="7196" id="7196"></a>
<p>Real Application Clusters environments are fully redundant because all nodes access all the disks in the cluster. The failure of one node does not affect another node's ability to process transactions. As long as the cluster has one surviving node, all database clients can process all transactions, although subject to increased response times due to capacity constraints on the one node.</p>
<a name="6930" id="6930"></a>
<h4><span color="#330099">Cluster Interconnects</span></h4>
<a name="6931" id="6931"></a>
<p>Interconnect redundancy is often overlooked in clustered systems. This is because the <a href="https://docs.oracle.com/cd/A91202_01/901_doc/rac.901/a89867/glossary.htm#432760"><strong>mean time to failure (MTTF)</strong></a> is generally several years. Therefore, cluster interconnect redundancy might not be a high priority. Also, depending on the system and sophistication level, a redundant cluster interconnect could be cost prohibitive and have insufficient business justification.</p>
<a name="6932" id="6932"></a>
<p>However, a redundant cluster interconnect is an important aspect of a fully redundant cluster. Without this, a system is not truly free of single points-of-failure. Cluster interconnects can fail for a variety of reasons and not all of them are accounted for. Nor can the be accounted for when manufacturer MTTF metrics are provided. Interconnects can fail due to either device malfunctions, such as an oscillator failure in a switch interconnect, or because of human error.</p>
<a name="6933" id="6933"></a>
<h4><span color="#330099">Storage Devices</span></h4>
<a name="7260" id="7260"></a>
<p>Real Application Clusters operate on a single image of the data; all nodes in the cluster access the same set of data files. Database administrators are encouraged to use hardware-based mirroring to maintain redundant media. In this regard, Real Application Clusters are no different from single instance Oracle. Disk redundancy depends on the underlying hardware and software mirroring in use, such as a <a href="https://docs.oracle.com/cd/A91202_01/901_doc/rac.901/a89867/glossary.htm#433295"><strong>Redundant Array of Independent Disks (RAID)</strong></a>.</p>
<a name="6936" id="6936"></a>
<h4><span color="#330099">Operating System Software and Cluster Managers</span></h4>
<a name="7812" id="7812"></a>
<p>Real Application Clusters environments have full node redundancy; each node runs its own operating system copy. Hence, the same considerations about node redundancy also apply to the operating system. The <a href="https://docs.oracle.com/cd/A91202_01/901_doc/rac.901/a89867/glossary.htm#432087"><strong>Cluster Manager (CM)</strong></a> is an extension of the operating system. Since the Cluster Manager software is also installed on all the nodes of the cluster, full redundancy is assured.</p>
<a name="6939" id="6939"></a>
<h4><span color="#330099">Database Software</span></h4>
<a name="6940" id="6940"></a>
<p>In Real Application Clusters, Oracle executables (such as Oracle home) are installed on the local disks of each node and an instance runs on each node of the cluster. Note that if your supports a cluster file system (CFS) then only one copy of Oracle home will be installed. All instances have equal access to all data and can process any transactions. In this way, Real Application Clusters ensure full database software redundancy.</p>
<a name="6941" id="6941"></a>
<h2><span color="#330099">Disaster Planning</span></h2>
<a name="12248" id="12248"></a>
<p>Real Application Clusters are primarily a single site, high availability solution. This means the nodes in the cluster generally exist within the same building, if not the same room. Thus, disaster planning can be critical. Disaster planning covers planning for fires, floods, hurricanes, earthquakes, terrorism, and so on. Depending on how mission critical your system is, and the propensity of your system's location for such disasters, disaster planning could be an important high availability component.</p>
<a name="6944" id="6944"></a>
<p>Oracle offers other solutions such as Oracle9<em>i</em> Data Guard and Oracle Replication to facilitate more comprehensive disaster recovery planning. You can use these solutions with Real Application Clusters where one cluster hosts the primary database and another remote system or cluster hosts the disaster recovery database. However, Real Application Clusters are not required on either site for purposes of disaster recovery.</p>
<a name="6945" id="6945"></a>
<h2><span color="#330099">Failure Protection Validation</span></h2>
<a name="7896" id="7896"></a>
<p>Once you have carefully considered your system level issues, validate that the Real Application Clusters environment optimally protects against potential failures. Use the following list of failure causes to plan and troubleshoot your failure protection system:</p>
<ul>
<li type="disc">
<a name="7321" id="7321"></a>Cluster component</li>
<li type="disc">
<a name="6948" id="6948"></a>CPU</li>
<li type="disc">
<a name="6949" id="6949"></a>Memory</li>
<li type="disc">
<a name="6950" id="6950"></a>Interconnect software</li>
<li type="disc">
<a name="6951" id="6951"></a>Operating System</li>
<li type="disc">
<a name="6952" id="6952"></a>Cluster Manager</li>
<li type="disc">
<a name="6953" id="6953"></a>Oracle database instance media</li>
<li type="disc">
<a name="6954" id="6954"></a>Corrupt or lost control file</li>
<li type="disc">
<a name="6955" id="6955"></a>Corrupt or lost log file</li>
<li type="disc">
<a name="6956" id="6956"></a>Corrupt or lost data file</li>
<li type="disc">
<a name="15197" id="15197"></a>Human error</li>
<li type="disc">
<a name="6957" id="6957"></a>Dropped or deleted database object</li>
</ul>
<a name="6958" id="6958"></a>
<p>Real Application Clusters environments protect against cluster component failures and software failures. However, media failures and human error could still cause system downtime. Real Application Clusters, as with single instance Oracle, operates on one set of files. For this reason, you should adopt best practices to avoid the adverse effects of media failures.</p>
<a name="8526" id="8526"></a>
<p>RAID-based redundancy practices avoid file loss but might not prevent rare cases of file corruptions. Also, if you mistakenly drop a database object in an Real Application Clusters environment, then you can recover that object the same way you would in a single instance database. These are the primary limitations in an otherwise very robust and highly available Real Application Clusters system.</p>
<a name="9876" id="9876"></a>
<p>Once you deploy your system, the key issue is the transparency of failover and its duration.</p>
<a name="15601" id="15601">
<div align="center">
<span><strong>See Also:</strong></span><a name="15609" id="15609"></a><a href="https://docs.oracle.com/cd/A91202_01/901_doc/rac.901/a89867/pshavdtl.htm#6915">"System Level Planning"</a>for more information about how Real Application Clusters environments protect against cluster component failures and software failures&nbsp;
<table border="0" cellpadding="0" width="80%" cellspacing="0" dir="ltr">
<tbody>
<tr>
<td></td>
</tr>
</tbody>
</table>
</div>
</a> <a name="13837" id="13837"></a>
<h2><span color="#330099">Failover and Real Application Clusters</span></h2>
<a name="13838" id="13838"></a>
<p>This section describes the basic principles of failover and the various features Real Application Clusters offer to implement it in high availability systems. Topics in this section include:</p>
<ul>
<li type="disc">
<a name="9892" id="9892"></a><a href="https://docs.oracle.com/cd/A91202_01/901_doc/rac.901/a89867/pshavdtl.htm#9371">Failover Basics</a>
</li>
<li type="disc">
<a name="9903" id="9903"></a><a href="https://docs.oracle.com/cd/A91202_01/901_doc/rac.901/a89867/pshavdtl.htm#10853">Client Failover</a>
</li>
<li type="disc">
<a name="16303" id="16303"></a><a href="https://docs.oracle.com/cd/A91202_01/901_doc/rac.901/a89867/pshavdtl.htm#6987">Uses of Transparent Application Failover</a>
</li>
<li type="disc">
<a name="9904" id="9904"></a><a href="https://docs.oracle.com/cd/A91202_01/901_doc/rac.901/a89867/pshavdtl.htm#7011">Server Failover</a>
</li>
</ul>
<a name="9371" id="9371"></a>
<h3><span color="#330099">Failover Basics</span></h3>
<a name="14593" id="14593"></a>
<p>Failover requires that highly available systems have accurate instance monitoring or <a href="https://docs.oracle.com/cd/A91202_01/901_doc/rac.901/a89867/glossary.htm#432497"><strong>heartbeat</strong></a> mechanisms. In addition to having this functionality for normal operations, the system must be able to quickly and accurately synchronize resources during failover.</p>
<a name="14596" id="14596"></a>
<p>The process of synchronizing, or <em>remastering</em>, requires the graceful shutdown of the failing system as well as an accurate assumption of control of the resources that were mastered on that system. Accurate remastering also requires that the system have adequate information about resources across the cluster. This means your system must record resource information to remote nodes as well as local. This makes the information needed for failover and recovery available to the recovering instances.</p>
<a name="13308" id="13308"></a> <a name="10122" id="10122"></a>
<h4><span color="#330099">Duration of Failover</span></h4>
<a name="10851" id="10851"></a>
<p>The duration of failover includes the time a system requires to remaster systemwide resources and recover from failures. The duration of the failover process can be a relatively short interval on certified platforms.</p>
<ul>
<li type="disc">
<a name="14988" id="14988"></a>For existing users, failover entails both server and client failover actions.</li>
<li type="disc">
<a name="14989" id="14989"></a>For new users, failover only entails the duration of server failover processing.</li>
</ul>
<a name="10853" id="10853"></a>
<h3><span color="#330099">Client Failover</span></h3>
<a name="6960" id="6960"></a>
<p>It is important to hide system failures from database client connections. Such connections can include application users in client server environments or middle-tier database clients in multitiered application environments. Properly configured failover mechanisms transparently reroute client sessions to an available node in the cluster. This capability in the Oracle database is referred to as Transparent Application Failover.</p>
<a name="6961" id="6961"></a>
<h4><span color="#330099">Transparent Application Failover</span></h4>
<a name="15664" id="15664"></a>
<p><a href="https://docs.oracle.com/cd/A91202_01/901_doc/rac.901/a89867/glossary.htm#433759"><strong>Transparent Application Failover (TAF)</strong></a> enables an application user to automatically reconnect to a database if the connection fails. Active transactions roll back, but the new database connection, made by way of a different node, is identical to the original. This is true regardless of how the connection fails.</p>
<a name="15665" id="15665"></a>
<h4><span color="#330099">How Transparent Application Failover Works</span></h4>
<a name="15666" id="15666"></a>
<p>With Transparent Application Failover, a client notices no loss of connection as long as there is one instance left serving the application. The <a href="https://docs.oracle.com/cd/A91202_01/901_doc/rac.901/a89867/glossary.htm#432159"><strong>database administrator (DBA)</strong></a> controls which applications run on which instances and also creates a failover order for each application.</p>
<a name="13004" id="13004"></a>
<h4><span color="#330099">Elements Affected by Transparent Application Failover</span></h4>
<a name="13005" id="13005"></a>
<p>During normal client/server database operations, the client maintains a connection to the database so the client and server can communicate. If the server fails, so then does the connection. The next time the client tries to use the connection the client issues an error. At this point, the user must log in to the database again.</p>
<a name="9979" id="9979"></a>
<p>With Transparent Application Failover, however, Oracle automatically obtains a new connection to the database. This enables users to continue working as if the original connection had never failed.</p>
<a name="6967" id="6967"></a>
<p>There are several elements associated with active database connections. These include:</p>
<ul>
<li type="disc">
<a name="6968" id="6968"></a>Client/Server database connections</li>
<li type="disc">
<a name="6969" id="6969"></a>Users' database sessions executing commands</li>
<li type="disc">
<a name="6970" id="6970"></a>Open cursors used for fetching</li>
<li type="disc">
<a name="6971" id="6971"></a>Active transactions</li>
<li type="disc">
<a name="6972" id="6972"></a>Server-side program variables</li>
</ul>
<a name="6973" id="6973"></a>
<p>Transparent Application Failover automatically restores some of these elements. However, you might need to embed other elements in the application code to enable transparent application failover.</p>
<a name="9993" id="9993"></a> <a name="6987" id="6987"></a>
<h3><span color="#330099">Uses of Transparent Application Failover</span></h3>
<a name="6988" id="6988"></a>
<p>While the ability to fail over client sessions is an important benefit of Transparent Application Failover, there are other useful scenarios where Transparent Application Failover improves system availability. These topics are discussed in the following subsections:</p>
<ul>
<li type="disc">
<a name="7441" id="7441"></a><a href="https://docs.oracle.com/cd/A91202_01/901_doc/rac.901/a89867/pshavdtl.htm#7444">Transactional Shutdowns</a>
</li>
<li type="disc">
<a name="16322" id="16322"></a><a href="https://docs.oracle.com/cd/A91202_01/901_doc/rac.901/a89867/pshavdtl.htm#16003">Quiescing the Database</a>
</li>
<li type="disc">
<a name="7442" id="7442"></a><a href="https://docs.oracle.com/cd/A91202_01/901_doc/rac.901/a89867/pshavdtl.htm#6994">Load Balancing</a>
</li>
<li type="disc">
<a name="16326" id="16326"></a><a href="https://docs.oracle.com/cd/A91202_01/901_doc/rac.901/a89867/pshavdtl.htm#6998">Transparent Application Failover Restrictions</a>
</li>
<li type="disc">
<a name="16331" id="16331"></a><a href="https://docs.oracle.com/cd/A91202_01/901_doc/rac.901/a89867/pshavdtl.htm#7001">Database Client Processing During Failover</a>
</li>
</ul>
<a name="7444" id="7444"></a>
<h4><span color="#330099">Transactional Shutdowns</span></h4>
<a name="15092" id="15092"></a>
<p>It is sometimes necessary to take nodes out of service for maintenance or repair. For example, if you want to apply patch releases without interrupting service to application clients. Transactional shutdowns facilitate shutting down selected nodes rather than an entire database. Two transactional shutdown options are available:</p>
<ul>
<li type="disc">
<a name="15102" id="15102"></a>Use the <code>TRANSACTIONAL</code> clause of the <code>SHUTDOWN</code> statement to take a node out of service so that the shutdown event is deferred until all existing transactions are completed. In this way client sessions can be migrated to another node of the cluster at transaction boundaries.</li>
<li type="disc">
<a name="16281" id="16281"></a>Use the <code>TRANSACTIONAL LOCAL</code> clause of the <code>SHUTDOWN</code> statement to do a transactional shutdown on a specified local instance. This command can be used to prevent new transactions from starting locally, and to perform an immediate shutdown after all local transactions have completed. With this option, you can gracefully move all sessions from one instance to another by shutting down selected instances transactionally.</li>
</ul>
<a name="6993" id="6993"></a>
<p>After performing a transactional shutdown, newly submitted transactions get routed to an alternate node in the cluster. An immediate shutdown is performed on the node when all existing transactions complete.</p>
<a name="16003" id="16003"></a>
<h4><span color="#330099">Quiescing the Database</span></h4>
<a name="16024" id="16024"></a>
<p>You may need to perform administrative tasks that require isolation from concurrent user transactions or queries. To do this, you can use the quiesce database feature. This prevents you, for example, from having to shut down the database and re-open it in restricted mode to perform such tasks.</p>
<a name="16025" id="16025"></a>
<p>To do this, you can use the <code>ALTER SYSTEM</code> statement with the <code>QUIESCE RESTRICTED</code> clause.</p>
<a name="16034" id="16034">
<div align="center">
<span><strong>Note:</strong></span><a name="16033" id="16033"></a>
<p>You cannot open the database on one instance if the database is <em>being quiesced</em> on another node. In other words, if you issued the <code>ALTER SYSTEM QUIESCE RESTRICTED</code> statement but it is not finished processing, you cannot open the database. Nor can you open the database if it is already in a quiesced state. &nbsp;</p>
<hr>
<table border="0" cellpadding="0" width="80%" cellspacing="0" dir="ltr">
<tbody>
<tr>
<td>
<hr>
</td>
</tr>
</tbody>
</table>
</div>
</a> <a name="16049" id="16049"></a> <a name="16006" id="16006"></a>
<p>The QUIESCE RESTRICTED clause enables you to perform administrative tasks in isolation from concurrent user transactions or queries.</p>
<a name="6994" id="6994"></a>
<h4><span color="#330099">Load Balancing</span></h4>
<a name="6995" id="6995"></a>
<p>A database is available when it processes transactions in a timely manner. When the load exceeds a node's capacity, client transaction response times are adversely affected and the database availability is compromised. It then becomes important to manually migrate a group of client sessions to a less heavily loaded node to maintain response times and application availability.</p>
<a name="15011" id="15011"></a>
<p>In Real Application Clusters, the Transport Network Services (TNS) listener provides automated load balancing across nodes in both shared server and dedicated server configurations.</p>
<a name="15177" id="15177"></a>
<p><strong>Connection load balancing</strong> improves connection performance by balancing the number of active connections among multiple dispatchers. In a single-instance environment, the listener selects the least loaded dispatcher to handle the incoming client requests. In a Real Application Clusters environment, connection load balancing also has the capability to balance the number of active connections among multiple instances.</p>
<a name="15162" id="15162"></a>
<p>Due to dynamic service registration, a listener is always aware of all of the instances and dispatchers regardless of their locations. Depending on the load information, a listener decides which instance and, if in shared server configuration, to which dispatcher to send the incoming client request.</p>
<a name="15163" id="15163"></a>
<p>In a shared server configuration, a listener selects a dispatcher in the following order:</p>
<ol type="1">
<li type="1" value="1">
<a name="15690" id="15690"></a>Least loaded node</li>
<li type="1" value="2">
<a name="15691" id="15691"></a>Least loaded instance</li>
<li type="1" value="3">
<a name="15692" id="15692"></a>Least loaded dispatcher for that instance</li>
</ol>
<a name="15693" id="15693"></a> <a name="15694" id="15694"></a>
<p>In a dedicated server configuration, a listener selects an instance in the following order:</p>
<ol type="1">
<li type="1" value="1">
<a name="15701" id="15701"></a>Least loaded node</li>
<li type="1" value="2">
<a name="15709" id="15709"></a>Least loaded instance</li>
</ol>
<a name="15164" id="15164"></a>
<p>If a database service has multiple instances on multiple nodes, then the listener chooses the least loaded instance on the least loaded node. If shared server is configured, then the least loaded dispatcher of the selected instance is chosen.</p>
<a name="15817" id="15817">.</a> <a name="6998" id="6998"></a>
<h4><span color="#330099">Transparent Application Failover Restrictions</span></h4>
<a name="7000" id="7000"></a>
<p>When a connection is lost, you might experience the following:</p>
<ul>
<li type="disc">
<a name="7474" id="7474"></a>All PL/SQL package states on the server are lost at failover</li>
<li type="disc">
<a name="7478" id="7478"></a>ALTER SESSION statements are lost</li>
<li type="disc">
<a name="7479" id="7479"></a>If failover occurs when a transaction is in process, then each subsequent call causes an error message until the user issues an OCITransRollback <a href="https://docs.oracle.com/cd/A91202_01/901_doc/rac.901/a89867/glossary.htm#432038"><strong>call</strong></a>. Then Oracle issues an <a href="https://docs.oracle.com/cd/A91202_01/901_doc/rac.901/a89867/glossary.htm#432926"><strong>Oracle Call Interface (OCI)</strong></a> success message. Be sure to check this message to see if you must perform additional operations.</li>
<li type="disc">
<a name="7480" id="7480"></a>Continuing work on failed-over cursors can cause an error message</li>
</ul>
<a name="7482" id="7482"></a>
<p>If the first command after failover is not a SQL SELECT or OCIStmtFetch statement, then an error message results. Failover only takes effect if the application is programmed with OCI release 8.0 or greater.</p>
<a name="7001" id="7001"></a>
<h4><span color="#330099">Database Client Processing During Failover</span></h4>
<a name="10866" id="10866"></a>
<p>Failover processing for query clients is different than the failover processing for Database Mount Lock clients. The important issue during failover operations in either case is the failure is masked from existing client connections as much as possible. The following subsections describe both types of failover processing.</p>
<a name="10867" id="10867"></a>
<h5><span color="#330099">Query Clients</span></h5>
<a name="7004" id="7004"></a>
<p>At failover, in-progress queries are reissued and processed from the beginning. This might extend the duration of the next query if the original query took a long time. With Transparent Application Failover (TAF), the failure is masked for query clients with an increased response time being the only client observation. If the client query can be satisfied with data in the buffer cache of the surviving node that the client reconnected to, then the increased response time is minimal. By using TAF's PRECONNECT method eliminates the need to reconnect to a surviving instance and thus further minimizes response time. However, <code>PRECONNECT</code> allocates resources awaiting the failover event.</p>
<a name="7005" id="7005"></a>
<p>If the client query cannot be satisfied with data in the buffer cache of the reconnect node, then disk I/O is necessary to process the client query. However, server-side recovery needs to complete before access to the data files is allowed. The client transaction experiences a system pause until server-side recovery completes, if server-side recovery has not already completed.</p>
<a name="10258" id="10258"></a>
<p>You can also use a callback function through an OCI call to notify clients of the failover so that the clients do not misinterpret the delay for a failure. This prevents the clients from manually attempting to reestablish their connections.</p>
<a name="7006" id="7006"></a>
<h5><span color="#330099">Database Mount Lock Clients</span></h5>
<a name="13535" id="13535"></a>
<p>Database Mount Lock database clients perform INSERT, UPDATE, and DELETE operations. In-progress Database Mount Lock transactions on a failed instance may restart on a surviving instance without client knowledge providing the application code fully exploits the OCI libraries. This achieves application failover, without manual reconnects, but requires application-level coding. The required code handles certain Oracle errors and performs a reconnect when those errors occur.</p>
<a name="7008" id="7008"></a>
<p>Without this application code, INSERT, UPDATE, and DELETE operations on the failed instance return an un-handled Oracle error code. In such cases, Oracle resubmits the transaction for execution. Upon re-submission, Oracle routes the client connections to a surviving instance. The client transaction then stops only momentarily until server-side recovery completes.</p>
<a name="7011" id="7011"></a>
<h3><span color="#330099">Server Failover</span></h3>
<a name="7012" id="7012"></a>
<p>Server-side failover processing in Real Application Clusters is different from host-based failover solutions that are available on many server platforms. The following subsections describe both types of failover processing.</p>
<a name="7013" id="7013"></a>
<h4><span color="#330099">Host-Based Failover</span></h4>
<a name="7014" id="7014"></a>
<p>Many operating system vendors and other cluster software vendors offer high availability application failover products. These failover solutions monitor application services on a given primary cluster node. They then fail over such services to a secondary cluster node as needed. Host-based failover solutions generally have one active instance performing useful work for a given database application. The secondary node monitors the application service on the primary node and initiates failover when the primary node service is unavailable.</p>
<a name="7015" id="7015"></a>
<p>Failover in host-based systems usually includes the following steps.</p>
<ol type="1">
<li type="1" value="1">
<a name="8190" id="8190"></a>Detecting failure by monitoring the heartbeat</li>
<li type="1" value="2">
<a name="8198" id="8198"></a>Reorganizing cluster membership in the Cluster Manager</li>
<li type="1" value="3">
<a name="8199" id="8199"></a>Transferring disk ownership from the primary node to a secondary node</li>
<li type="1" value="4">
<a name="8200" id="8200"></a>Restarting application and database binaries (Oracle executables)</li>
<li type="1" value="5">
<a name="8201" id="8201"></a>Performing application and database recovery</li>
<li type="1" value="6">
<a name="8202" id="8202"></a>Reestablishing client connections to the failover node</li>
</ol>
<a name="13867" id="13867"></a>
<h4><span color="#330099">Real Application Clusters Failover</span></h4>
<a name="13868" id="13868"></a>
<p>Real Application Clusters provide rapid server-side failover. This is accomplished by the concurrent, active-active architecture in Real Application Clusters. In other words, multiple Oracle instances are concurrently active on multiple nodes and these instances synchronize access to the same database. All nodes also have concurrent ownership and access to all disks. When one node fails, all other nodes in the cluster maintain access to all the disks; there is no disk ownership to transfer, and database application binaries are already loaded into memory.</p>
<a name="7018" id="7018"></a>
<p>Depending on the size of the database, the duration of failover can vary. The larger the database, or the greater the size of its data files, the greater the failover benefit in using Real Application Clusters.</p>
<a name="15337" id="15337"></a>
<h2><span color="#330099">How Failover Works</span></h2>
<a name="15340" id="15340"></a>
<p>The following subsections describe server failover recovery processing in Real Application Clusters:</p>
<ul>
<li type="disc">
<a name="15723" id="15723"></a><a href="https://docs.oracle.com/cd/A91202_01/901_doc/rac.901/a89867/pshavdtl.htm#7025">Detecting Failure</a>
</li>
<li type="disc">
<a name="15728" id="15728"></a><a href="https://docs.oracle.com/cd/A91202_01/901_doc/rac.901/a89867/pshavdtl.htm#7027">Reorganizing Cluster Membership</a>
</li>
<li type="disc">
<a name="15733" id="15733"></a><a href="https://docs.oracle.com/cd/A91202_01/901_doc/rac.901/a89867/pshavdtl.htm#16161">Performing Database Recovery</a>
</li>
</ul>
<a name="7025" id="7025"></a>
<h3><span color="#330099">Detecting Failure</span></h3>
<a name="8259" id="8259"></a>
<p>Real Application Clusters relies on the Cluster Manager software for failure detection because the Cluster Manager maintains the heartbeat functions. The time it takes for the Cluster Manager to detect that a node is no longer in operation is a function of a configurable heartbeat timeout parameter. This parameter varies, depending on your platform. You can configure this value on most systems. Defaults can vary significantly, depending on your clusterware. (Such as Sun Cluster or Hewlett-Packard Service Guard OPS Edition.) The parameter value is inversely related to the number of false failure detections because the cluster might incorrectly determine that a node is failing due to transient failures if the timeout interval is set too low. When a failure is detected, cluster reorganization occurs.</p>
<a name="7027" id="7027"></a>
<h3><span color="#330099">Reorganizing Cluster Membership</span></h3>
<a name="13491" id="13491"></a>
<p>When a node fails, Oracle must alter the node's cluster membership status. This is known as a <em>cluster reorganization</em> and it usually happens quickly. The duration of cluster reorganization is proportional to the number of surviving nodes in the cluster.</p>
<a name="13493" id="13493"></a>
<p>The <a href="https://docs.oracle.com/cd/A91202_01/901_doc/rac.901/a89867/glossary.htm#432387"><strong>Global Cache Service (GCS)</strong></a> and <a href="https://docs.oracle.com/cd/A91202_01/901_doc/rac.901/a89867/glossary.htm#432414"><strong>Global Enqueue Service (GES)</strong></a> provide the Cluster Manager interfaces to the software and expose the cluster membership map to the Oracle instances when nodes are added or deleted from the cluster. The LMON process on each cluster node communicates with the Cluster Manager on the respective nodes and exposes that information to the respective Oracle instances.</p>
<a name="12505" id="12505"></a>
<p>LMON also provides another useful function by continually sending messages from the node it runs on and often writing to the shared disk. The absence of these functions from any node is evidence to the surviving nodes that the node is no longer a member of the cluster. Such a failure causes a change in a node's membership status within the cluster and LMON initiates the recovery actions that include remastering of Global Cache Service and Global Enqueue Service resources and instance recovery.</p>
<a name="12506" id="12506"></a>
<p>At this stage, the Real Application Clusters environment is in a state of system pause, and most client transactions suspend until Oracle completes recovery processing.</p>
<a name="16159" id="16159"></a> <a name="16171" id="16171"></a>
<h5><span color="#330099">Instance Membership Recovery</span></h5>
<a name="16175" id="16175"></a>
<p>The process of <a href="https://docs.oracle.com/cd/A91202_01/901_doc/rac.901/a89867/glossary.htm#432572"><strong>instance membership recovery (IMR)</strong></a> guarantees that all members of a cluster are functional (<a href="https://docs.oracle.com/cd/A91202_01/901_doc/rac.901/a89867/glossary.htm#434531"><strong>active</strong></a>) The IMR process does the following:</p>
<ul>
<li type="disc">
<a name="16179" id="16179"></a>Ensures that communications are viable between all members, and that all members are capable of responding.</li>
<li type="disc">
<a name="16180" id="16180"></a>Provides a mechanism for removing members that are not active. IMR arbitrates the membership and removes members that it decides no longer belong to the cluster.</li>
<li type="disc">
<a name="16182" id="16182"></a>Votes on membership using the Control File. Each member writes a bitmap to the Control File. This is part of the checkpoint progress record.</li>
<li type="disc">
<a name="16183" id="16183"></a>Removes members based on a communication failure. IMR will perceive members as <em>expired</em> if they do not provide their normal periodic heartbeat to the Control File, or if they do not respond to a query on their status.</li>
<li type="disc">
<a name="16187" id="16187"></a>Settles the membership votes and locks the <a href="https://docs.oracle.com/cd/A91202_01/901_doc/rac.901/a89867/glossary.htm#432121"><strong>control file voting results record (CFVRR)</strong></a> with an arbiter.</li>
</ul>
<a name="16189" id="16189"></a>
<p>All instances read the CFVRR. If a member is not in the membership map, then IMR assumes a node has expired. Appropriate diagnostic information is provided. As IMR is currently configured, all members wait indefinitely for notification of node expiration. There is no forced removal of instances.</p>
<a name="16193" id="16193"></a>
<p>Part of the fault tolerance of Real Application Clusters is provision for the possibility that the IMR arbiter itself could fail.</p>
<a name="16161" id="16161"></a>
<h3><span color="#330099">Performing Database Recovery</span></h3>
<a name="16162" id="16162"></a>
<p>When an instance fails, Oracle must remaster Global Cache Service resources from the failed instance onto the surviving cluster nodes. The database recovery process in Real Application Clusters includes these topics discussed in the following sections:</p>
<ul>
<li type="disc">
<a name="15752" id="15752"></a><a href="https://docs.oracle.com/cd/A91202_01/901_doc/rac.901/a89867/pshavdtl.htm#15424">Remastering Global Cache Service Resources of the Failed Instance</a>
</li>
<li type="disc">
<a name="15767" id="15767"></a><a href="https://docs.oracle.com/cd/A91202_01/901_doc/rac.901/a89867/pshavdtl.htm#7040">Instance Recovery</a>
</li>
</ul>
<a name="15424" id="15424"></a>
<h4><span color="#330099">Remastering Global Cache Service Resources of the Failed Instance</span></h4>
<a name="15427" id="15427"></a>
<p>The time required for remastering locks is proportional to the number of Global Cache Service resources in the failed instance. This number in turn depends upon the size of the buffer caches. These processes occur in parallel.</p>
<a name="7038" id="7038"></a>
<p>During this phase, all resources previously mastered at the failed instance are redistributed across the remaining instances. These resources are reconstructed at their new master instance. All other resources previously mastered at surviving instances are not affected. For any lock request, there is a 1/<em>n</em> chance that the request will be satisfied locally and a (<em>n</em>-1)/<em>n</em> chance that the lock request involves remote operations. In the case of one surviving instance, all lock operations are satisfied locally.</p>
<a name="7039" id="7039"></a>
<p>Once remastering of the failed instance Global Cache Service resource is completed, Oracle must clean up the in-progress transactions of the failed instance. This is known as <em>instance recovery</em>.</p>
<a name="7040" id="7040"></a>
<h4><span color="#330099">Instance Recovery</span></h4>
<a name="7041" id="7041"></a>
<p>Instance recovery includes cache recovery and transaction recovery. Instance recovery requires that an active Real Application Clusters instance detects failure and performs recovery processing for the failed instance. The first Real Application Clusters instance that detects the failure, by way of its LMON process, controls the recovery of the failed instance by taking over its redo log files and performing instance recovery. This is why the redo log files must be on a shared device such as a shared raw logical volume or cluster file system.</p>
<a name="7042" id="7042"></a>
<p>Instance recovery is complete when Oracle has performed the following steps:</p>
<ul>
<li type="disc">
<a name="12607" id="12607"></a>Rolling back all uncommitted transactions of the failed instance. This is also known as <em>transaction recovery</em>.</li>
<li type="disc">
<a name="16346" id="16346"></a>Replaying the online redo log files of the failed instance.</li>
</ul>
<a name="12661" id="12661"></a>
<p>Since Oracle can perform transaction recovery in a deferred fashion, client transactions can begin processing when cache recovery is complete.</p>
<a name="13271" id="13271"></a>
<p>You do not need to recover entire files in the event of a failure. You can recover individual blocks.</p>
<a name="15829" id="15829">.</a> <a name="7043" id="7043"></a>
<h5><span color="#330099">Cache Recovery</span></h5>
<a name="7044" id="7044"></a>
<p>For cache recovery, Oracle replays the online redo logs of the failed instance. Oracle performs cache recovery using parallel execution so that parallel processes, or threads, replay the redo logs of the failed Oracle instance. It could be important that you keep the time interval for redo log replay to a predictable duration. The Fast-Start Recovery feature in Oracle9<em>i</em> enables you to control this.</p>
<a name="8831" id="8831"></a>
<p>Oracle also provides nonblocking rollback capabilities. This means that full database access can begin as soon as Oracle has replayed the online log files. After cache recovery completes, Oracle begins transaction recovery.</p>
<a name="12741" id="12741"></a> <a name="7046" id="7046"></a>
<h5><span color="#330099">Transaction Recovery</span></h5>
<a name="7047" id="7047"></a>
<p>Transaction recovery comprises rolling back all uncommitted transactions of the failed instance. Uncommitted transactions are <em>in-progress</em> transactions that did not commit.</p>
<a name="7049" id="7049"></a>
<p>The Oracle9<em>i</em> Fast-Start Rollback feature performs this as deferred processing that runs in the background. Oracle uses a multi-version read consistency technology to provide on-demand rollback of only those rows blocked by expired transactions. This allows new transactions to progress with minimal delay. New transactions do not have to wait for long-running expired transactions to be rolled back. Therefore, large transactions generally do not affect database recovery time.</p>
<a name="7050" id="7050"></a>
<p>Just as with cache recovery, Oracle9<em>i</em> Fast-Start Rollback rolls back expired transactions in parallel. However, single instance Oracle rolls back expired transactions by using the CPU of one node.</p>
<a name="7051" id="7051"></a>
<p>Real Application Clusters provide cluster-aware Fast-Start Rollback capabilities that use all the CPU nodes of a cluster to perform parallel rollback operations. Each cluster node spawns a recovery coordinator and recovery processes to assist with parallel rollback operations. The Fast-Start Rollback feature is thus <em>cluster aware</em> because the database is aware of and utilizes all cluster resources for parallel rollback operations.</p>
<a name="7052" id="7052"></a>
<p>While the default behavior is to defer transaction recovery, you could choose to configure your system so transaction recovery completes before allowing client transactions to progress. In this scenario, the ability of Real Application Clusters to parallelize transaction recovery across multiple nodes is a more visible user benefit.</p>
<a name="10874" id="10874"></a>
<h2><span color="#330099">High Availability Configurations</span></h2>
<a name="11130" id="11130"></a>
<p>This section discusses these Real Application Clusters high availability configurations:</p>
<ul>
<li type="disc">
<a name="11134" id="11134"></a><a href="https://docs.oracle.com/cd/A91202_01/901_doc/rac.901/a89867/pshavdtl.htm#11144">Default n-node Configuration</a>
</li>
<li type="disc">
<a name="11138" id="11138"></a><a href="https://docs.oracle.com/cd/A91202_01/901_doc/rac.901/a89867/pshavdtl.htm#10876">Basic High Availability Configurations</a>
</li>
<li type="disc">
<a name="11142" id="11142"></a><a href="https://docs.oracle.com/cd/A91202_01/901_doc/rac.901/a89867/pshavdtl.htm#10886">Shared High Availability Node Configuration</a>
</li>
</ul>
<a name="11144" id="11144"></a>
<h3><span color="#330099">Default <em>n</em>-node Configuration</span></h3>
<a name="10906" id="10906"></a>
<p>The Real Application Clusters <em>n</em>-node configuration is the default environment. All nodes of the cluster participate in client transaction processing and client sessions can be load balanced at connect time. Response time is optimized for available cluster resources, such as CPU and memory, by distributing the load across cluster nodes to create a highly available environment.</p>
<a name="10875" id="10875"></a>
<h4><span color="#330099">Benefits of <em>n</em>-Node Configurations</span></h4>
<a name="10907" id="10907"></a>
<p>In the event of node failures, an instance on another node performs the necessary recovery actions. The database clients on the failed instance can be load balanced across the surviving (<em>n</em>-1) instances of the cluster. The increased load on each of the surviving instances can be minimized and availability increased by keeping response times within acceptable bounds. In this configuration, the database application workload can be distributed across all nodes and therefore provide optimal utilization of cluster machine resources.</p>
<a name="10876" id="10876"></a>
<h3><span color="#330099">Basic High Availability Configurations</span></h3>
<a name="10909" id="10909"></a>
<p>You can easily configure a basic high availability system for Real Application Clusters in two-node environments. The primary instance on one node accepts user connections while the <a href="https://docs.oracle.com/cd/A91202_01/901_doc/rac.901/a89867/glossary.htm#433442"><strong>secondary instance</strong></a> on the other node accepts connections when the primary node fails, or when specifically selected through the <code>INSTANCE_ROLE parameter</code>. You can configure this manually by controlling the routing of transactions to specific instances. However, Real Application Clusters provides the <a href="https://docs.oracle.com/cd/A91202_01/901_doc/rac.901/a89867/glossary.htm#433169"><strong>Primary/Secondary Configuration</strong></a> feature to accomplish this automatically.</p>
<a name="10877" id="10877"></a>
<p>Configure the Primary/Secondary Instance feature by setting the <a href="https://docs.oracle.com/cd/A91202_01/901_doc/rac.901/a89867/glossary.htm#432531"><strong>initsid.ora</strong></a> parameter <code>ACTIVE_INSTANCE_COUNT</code> to <code>1</code>. In the two-node environment, the instance that first mounts the database assumes the <a href="https://docs.oracle.com/cd/A91202_01/901_doc/rac.901/a89867/glossary.htm#433154"><strong>primary instance role</strong></a>. The other instance assumes the role of <a href="https://docs.oracle.com/cd/A91202_01/901_doc/rac.901/a89867/glossary.htm#433442"><strong>secondary instance</strong></a>. If the <a href="https://docs.oracle.com/cd/A91202_01/901_doc/rac.901/a89867/glossary.htm#433145"><strong>primary instance</strong></a> fails, then the secondary instance assumes the primary role. When the failed instance returns to active status, it assumes the secondary instance role.</p>
<a name="14990" id="14990"></a>
<h4><span color="#330099">Remote Clients</span></h4>
<a name="14721" id="14721"></a>
<p>The secondary instance becomes the primary instance only after the Cluster Manager informs it about the failure of the primary instance but before Global Cache Service and Global Enqueue Service reconfiguration and cache and transaction recovery processes begin. The redirection to the surviving instance happens transparently; application programming is not required. You only need to make minor configuration changes to the client connect strings.</p>
<a name="12079" id="12079"></a>
<p>In the Primary/Secondary Configuration, both instances run concurrently, like in any <em>n</em>-node Real Application Clusters environment. However, database application users only connect to the designated primary instance. The primary node masters all of the Global Cache Service and Global Enqueue Service resources. This minimizes communication between the nodes and provides performance levels that are nearly comparable to a traditional single node database.</p>
<a name="11771" id="11771"></a>
<p>The secondary instance can be used by specially configured clients, known as administrative clients, for batch query reporting operations or database administration tasks. This enables some level of utilization of the second node. It might also help off-load CPU capacity from the primary instance and justify the investment in redundant nodes.</p>
<a name="15903" id="15903"></a>
<p>The Primary/Secondary Configuration feature works in both dedicated server and shared server environments. However, it functions differently in each as described in the following sections:</p>
<a name="11809" id="11809"></a>
<h4><span color="#330099">Primary/Secondary Instance in Dedicated Server Environments</span></h4>
<a name="12031" id="12031"></a>
<p>In current high availability configurations, dedicated server environments do not use cross-instance listener registration. Connection requests made to a specific instance's listener can only be connected to that instance's service. This behavior is similar to the default <em>n</em>-node configuration in dedicated server environments.</p>
<a name="14395" id="14395"></a>
<p><a href="https://docs.oracle.com/cd/A91202_01/901_doc/rac.901/a89867/pshavdtl.htm#12036">Figure&nbsp;10-3</a> shows a cluster configuration before a node failure.</p>
<ol type="1">
<ol type="1">
<li type="1" value="1">
<a name="14338" id="14338"></a>SALES1 is in contact with a listener.</li>
<li type="1" value="2">
<a name="14344" id="14344"></a>A client is in contact with a listener.</li>
<li type="1" value="3">
<a name="14342" id="14342"></a>SALES1 becomes the primary instance.</li>
<li type="1" value="4">
<a name="15943" id="15943"></a>Soon after, SALES2 will become the secondary instance.</li>
</ol>
</ol>
<a name="12036" id="12036"></a>
<h4><span><em>Figure 10-3 Primary/Secondary Instance Feature in Dedicated Server Environments</em></span></h4>
<br>
<a name="12172" id="12172"><img width="600" alt="Text description of pss81016.gif follows" src="https://docs.oracle.com/cd/A91202_01/901_doc/rac.901/a89867/pss81016.gif" height="368"></a><a href="https://docs.oracle.com/cd/A91202_01/901_doc/rac.901/a89867/img_text/pss81016.htm"><br>
Text description of the illustration pss81016.gif</a> <a name="12054" id="12054"></a>
<p>When the primary instance fails, as shown in <a href="https://docs.oracle.com/cd/A91202_01/901_doc/rac.901/a89867/pshavdtl.htm#14535">Figure&nbsp;10-4</a>, the following steps occurs:</p>
<ol>
<ol type="1">
<li type="1" value="1">
<a name="14347" id="14347"></a>The failure of SALES1 is noted.</li>
<li type="1" value="2">
<a name="14349" id="14349"></a>A reconnection request from the client is rejected by the failed instance's listener.</li>
<li type="1" value="3">
<a name="14348" id="14348"></a>The secondary instance performs recovery and becomes the primary instance.</li>
<li type="1" value="4">
<a name="14533" id="14533"></a>Upon resubmitting the client request, the client reestablishes the connection by using the new primary instance's listener that then connects the client to the new primary instance. Note that the connection is reestablished automatically when you use address lists or if your client is configured to use connection failover.</li>
</ol>
</ol>
<a name="14535" id="14535"></a>
<h4><span><em>Figure 10-4 Node Failure in Dedicated Server Environments</em></span></h4>
<br>
<a name="11844" id="11844"><img width="600" alt="Text description of pss81017.gif follows" src="https://docs.oracle.com/cd/A91202_01/901_doc/rac.901/a89867/pss81017.gif" height="320"></a><a href="https://docs.oracle.com/cd/A91202_01/901_doc/rac.901/a89867/img_text/pss81017.htm"><br>
Text description of the illustration pss81017.gif</a> <a name="11248" id="11248"></a>
<h4><span color="#330099">Primary/Secondary Instance Feature and the Shared Server</span></h4>
<a name="11872" id="11872"></a>
<p>Real Application Clusters provides reconnection performance benefits when running in shared server mode. This is accomplished by the cross-registration of all the dispatchers and listeners in the cluster.</p>
<a name="12016" id="12016"></a>
<p>In the Primary/Secondary configurations, the primary instance's dispatcher registers as the primary instance with both listeners, as shown in <a href="https://docs.oracle.com/cd/A91202_01/901_doc/rac.901/a89867/pshavdtl.htm#11261">Figure&nbsp;10-5</a>:</p>
<ul>
<li type="disc">
<a name="14351" id="14351"></a>A client could connect to either listener. (Only the connection to the primary node listener is illustrated).</li>
<li type="disc">
<a name="14359" id="14359"></a>The client contacts the listener.</li>
<li type="disc">
<a name="14353" id="14353"></a>The listener then connects the client to the dispatcher. (Only the listener/dispatcher connection on the primary node is illustrated).</li>
</ul>
<a name="15934" id="15934"></a> <a name="11261" id="11261"></a>
<h4><span><em>Figure 10-5 Primary/Secondary Instance Feature in Shared Server Environments</em></span></h4>
<br>
<a name="12173" id="12173"><img width="600" alt="Text description of pss81019.gif follows" src="https://docs.oracle.com/cd/A91202_01/901_doc/rac.901/a89867/pss81019.gif" height="389"></a><a href="https://docs.oracle.com/cd/A91202_01/901_doc/rac.901/a89867/img_text/pss81019.htm"><br>
Text description of the illustration pss81019.gif</a> <a name="11953" id="11953"></a>
<p>Specially configured clients can use the secondary instance for batch operations. For example, batch reporting tasks or index creation operations can be performed on the secondary instance.</p>
<a name="11461" id="11461"></a> <a name="11503" id="11503"></a>
<p><a href="https://docs.oracle.com/cd/A91202_01/901_doc/rac.901/a89867/pshavdtl.htm#11270">Figure&nbsp;10-6</a> shows how a failed primary instance is replaced by a <em>new</em> primary instance.</p>
<ol type="1">
<li type="1" value="1">
<a name="14375" id="14375"></a>If the primary node fails, then the dispatcher in the secondary instance registers as the <em>new</em> primary instance with the listeners.</li>
<li type="1" value="2">
<a name="14383" id="14383"></a>The client requests a reconnection to the database through either listener.</li>
<li type="1" value="3">
<a name="14386" id="14386"></a>The listener directs the request to the new primary instance's dispatcher.</li>
</ol>
<a name="11270" id="11270"></a>
<h4><span><em>Figure 10-6 Node Failure in Shared Server Environments</em></span></h4>
<br>
<a name="12180" id="12180"><img width="600" alt="Text description of pss81018.gif follows" src="https://docs.oracle.com/cd/A91202_01/901_doc/rac.901/a89867/pss81018.gif" height="358"></a><a href="https://docs.oracle.com/cd/A91202_01/901_doc/rac.901/a89867/img_text/pss81018.htm"><br>
Text description of the illustration pss81018.gif</a> <a name="16210" id="16210"></a>
<h4><span color="#330099">Warming the Library Cache on the Secondary Instance</span></h4>
<a name="16211" id="16211"></a>
<p>Maintaining information about frequently executed SQL and PL/SQL statements in the library cache improves the performance of the Oracle database server. In Real Application Clusters primary and secondary instance configurations, the library cache associated with the primary instance contains up-to-date information. If failover occurs, then the benefit of that information is lost unless the library cache on the secondary instance is populated beforehand.</p>
<a name="16212" id="16212"></a>
<p>Use the <code>DBMS_LIBCACHE</code> package to transfer information in the library cache of the primary instance to the library cache of the secondary instance. This process is called <a href="https://docs.oracle.com/cd/A91202_01/901_doc/rac.901/a89867/glossary.htm#433814"><strong>warming the library cache</strong></a>. It improves performance immediately after failover because the new primary library cache does not need to be populated with parsed SQL statements and compiled PL/SQL units.</p>
<a name="16230" id="16230"></a> <a name="16208" id="16208"></a>
<h4><span color="#330099">Benefits of Basic High Availability Configurations</span></h4>
<a name="10910" id="10910"></a>
<p>The Primary/Secondary Instance feature provides two important reasons for using this scenario instead of a default two-node configuration. The Primary/Secondary Instance feature provides:</p>
<ul>
<li type="disc">
<a name="10913" id="10913"></a>A viable transition path for upgrading to an <em>n</em>-node configuration</li>
<li type="disc">
<a name="10915" id="10915"></a>A highly available solution for applications that do not need to scale beyond one node</li>
<li type="disc">
<a name="14328" id="14328"></a>Performance that is comparable to single instance databases</li>
<li type="disc">
<a name="14329" id="14329"></a>High performance without requiring tuning</li>
</ul>
<a name="10881" id="10881"></a>
<h4><span color="#330099">Transition Path to <em>n</em>-node Configurations</span></h4>
<a name="10916" id="10916"></a>
<p>The Primary/Secondary Instance feature offers a gradual way to migrate from a single instance application environment to a Real Application Clusters environment. This configuration also minimizes tuning issues because all client transactions are performed on one node at any given time. It also simplifies troubleshooting system problems because you tune only one node at a time as opposed to simultaneously tuning two or more nodes.</p>
<a name="10882" id="10882"></a>
<p>Since availability is also dependent on the DBA's ability to manage, tune, and troubleshoot the environment, the Real Application Clusters Primary/Secondary Instance feature provides a gradual way to ease the database administration staff into using Real Application Clusters.</p>
<a name="10886" id="10886"></a>
<h3><span color="#330099">Shared High Availability Node Configuration</span></h3>
<a name="10923" id="10923"></a>
<p>Running Real Application Clusters in an <em>n</em>-node configuration optimally utilizes the cluster resources. However, as discussed previously, this is not always possible or advisable. On the other hand, the financial investment required to have an idle node for failover is often prohibitive. These situations might instead be best suited for a shared high availability node configuration.</p>
<a name="10887" id="10887"></a>
<p>This type of configuration typically has several nodes each running a separate application module or service where all application services share one Real Application Clusters database. You can configure a separate designated node as a failover node. While an instance is running on that node, no users are being directed to it during normal operation. In the event that any one of the application nodes fails, Oracle can direct the workload to the failover node.</p>
<a name="10888" id="10888"></a>
<p>While this configuration is a useful one to consider for applications that need to run on separate nodes, it works best if a middle tier application or transaction processing monitor directs the appropriate application users to the appropriate nodes. Unlike the Primary/Secondary Instance Configuration, there is no database setup that automates the workload transition to the failover node. The application, or middle-tier software, would need to direct users from the failed application node to the designated failover node. The application would also need to control failing back the users once the failed node is operational. Failing back frees the failover node for processing user work from subsequent node failures.</p>
<a name="10889" id="10889"></a>
<h4><span color="#330099">Benefits of Shared High Availability Node Configurations</span></h4>
<a name="10924" id="10924"></a>
<p>In this configuration, application performance is maintained in the event of a failover. In the <em>n</em>-node configuration, application performance could degrade by 1/<em>n</em> due to the same workload being redistributed over a smaller set of cluster nodes.</p>
<a name="10556" id="10556"></a>
<h2><span color="#330099">Toward Deploying High Availability</span></h2>
<a name="10557" id="10557"></a>
<p>Real Application Clusters on clustered systems provides a fully redundant environment that is extremely fault resilient. Architecture is central to high availability in Real Application Clusters environments. All cluster nodes have an active instance that has equal access to all the data. If a node fails, then all users have access to all data by way of surviving instances on the other nodes. In-progress transactions on the failed node are recovered by the first node that detects the failure. In this way, there is minimal interruption to end user application availability with Real Application Clusters.</p>
</div>
</div>
<br>
<br>
<br>
<br>
<br>
<br>
via docs.oracle.com https://docs.oracle.com<br>
<br>
May 5, 2020 at 08:15AM
